// @gabrielcsapo/docusaurus-plugin-matomo/dist/index.js
"use strict";
exports.__esModule = true;
var path = require("path");
module.exports = function (context, options) {
    if (!options) {
        throw new Error("You need to specify options arguments use docusaurus-plugin-matomo.");
    }
    var siteUrl = options.siteUrl, matomoUrl = options.matomoUrl, _a = options.matomoPhpScript, matomoPhpScript = _a === void 0 ? "piwik.php" : _a, _b = options.matomoJsScript, matomoJsScript = _b === void 0 ? "piwik.js" : _b, siteId = options.siteId, dev = options.dev, localScript = options.localScript, requireConsent = options.requireConsent, requireCookieConsent = options.requireCookieConsent, disableCookies = options.disableCookies, cookieDomain = options.cookieDomain, enableJSErrorTracking = options.enableJSErrorTracking, _c = options.respectDnt, respectDnt = _c === void 0 ? true : _c;
    if (!matomoUrl || !siteId || !siteUrl) {
        throw new Error('You specified the "matomo" object in "themeConfig" but the "matomoUrl" or "siteUrl" or "siteId" field was missing. ' +
            "Please ensure this is not a mistake.");
    }
    var isProd = process.env.NODE_ENV === "production";
    var script = localScript ? localScript : "".concat(matomoUrl, "/").concat(matomoJsScript);
    var dntCondition = respectDnt
        ? "!(navigator.doNotTrack === '1' || window.doNotTrack === '1')"
        : "true";
    return {
        name: "docusaurus-plugin-matomo",
        getClientModules: function () {
            return isProd ? [path.resolve(__dirname, "./piwik.js")] : [];
        },
        injectHtmlTags: function () {
            if (!isProd) {
                return {};
            }
            return {
                headTags: [
                    {
                        tagName: "link",
                        attributes: {
                            rel: "preconnect",
                            href: script
                        }
                    },
                    {
                        tagName: "noscript",
                        innerHTML: "\n            var img = document.createElement('img');\n            img.src = \"".concat(matomoUrl, "/").concat(matomoPhpScript, "?idsite=").concat(siteId, "&rec=1&url=").concat(siteUrl, "\" + location.pathname;\n            img.style = \"border:0\";\n            img.alt = \"tracker\";\n\n            var s = document.getElementsByTagName('script')[0];\n            s.parentNode.insertBefore(img,s);\n            ")
                    },
                    {
                        tagName: "script",
                        innerHTML: "\n            window.dev = ".concat(dev, "\n            if (window.dev === true || ").concat(dntCondition, ") {\n              window._paq = window._paq || [];\n              ").concat(requireConsent ? "window._paq.push(['requireConsent']);" : "", "\n              ").concat(requireCookieConsent
                            ? "window._paq.push(['requireCookieConsent']);"
                            : "", "\n              ").concat(disableCookies ? "window._paq.push(['disableCookies']);" : "", "\n              ").concat(enableJSErrorTracking
                            ? "window._paq.push(['enableJSErrorTracking']);"
                            : "", "\n              ").concat(cookieDomain
                            ? "window._paq.push(['setCookieDomain', '".concat(cookieDomain, "']);")
                            : "", "\n              window._paq.push(['setTrackerUrl', '").concat(matomoUrl, "/").concat(matomoPhpScript, "']);\n              window._paq.push(['setSiteId', '").concat(siteId, "']);\n              window._paq.push(['enableHeartBeatTimer']);\n              window.start = new Date();\n              (function() {\n                var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];\n                g.type='text/javascript'; g.async=true; g.defer=true; g.src='").concat(script, "'; s.parentNode.insertBefore(g,s);\n              })();\n              if (window.dev === true) {\n                console.debug('[Matomo] Tracking initialized')\n                console.debug('[Matomo] matomoUrl: ").concat(matomoUrl, ", siteId: ").concat(siteId, "')\n              }\n            }\n            ")
                    },
                ]
            };
        }
    };
};
